dali/python/nvidia/dali/worker.py

```diff
@@ -0,0 +1,221 @@
+        return cls(scheduled, mem_chunk, data_batch)
+
+    @classmethod
+    def failed(cls, scheduled, exception, traceback_str=None):
+        return cls(scheduled, exception=exception, traceback_str=traceback_str)
+
+    def is_failed(self):
+        return self.exception is not None
+
+
+class SharedBatchesDispatcher:
+    """SharedBatchesDispatcher serializes batches, puts them into provided
+    shared memory chunks and notifies parent process of batch ready to be read from shared memory.
+    It keeps track of what shared memory chunks have been already sent and if needed, sends
+    file descriptors of memory chunks that parent process hasn't seen yet.
+
+    Parameters
+    ----------
+    `worker_id` : int
+        Id of the worker passed by the parent process. Added to messages sent over the ``res_pipe``
+        to simplify bookeeping in parent process.
+    `sock` : socket
+        Python wrapper around Unix socket, capable of sending file descriptors between processes.
+    `res_pipe`: pipe
+        Pipe used to send parent process a notification (along with essential meta data info) about
+        ready batch in a given shared memory chunk.
+    """
+
+    def __init__(self, worker_id, sock, res_pipe):
+        self.worker_id = worker_id
+        self.fd_sent = set()
+        self.sock = sock
+        self.res_pipe = res_pipe
+
+    def send(self, processed_tasks: _ProcessedTasks):
+        if processed_tasks.is_failed():  # one of the tasks failed
+            completed_tasks = CompletedTasks.failed(self.worker_id, processed_tasks)
+            self.res_pipe.send(completed_tasks)
+            return
+        mem_chunk = processed_tasks.mem_chunk
+        writer = SharedBatchWriter(processed_tasks.mem_chunk)
+        writer.write_batch(processed_tasks.data_batch)
+        batch_serialized = SharedBatchMeta.from_writer(writer)
+        completed_tasks = CompletedTasks.done(self.worker_id, processed_tasks, batch_serialized)
+        self.res_pipe.send(completed_tasks)
+        # send file descriptor for underlaying shared memory chunk if it hasn't been sent ever before
+        mem_chunk_id = batch_serialized.mem_chunk_id
+        if mem_chunk_id not in self.fd_sent:
+            self.fd_sent.add(mem_chunk_id)
+            reduction.sendfds(self.sock, [mem_chunk.shm_chunk.fd])
+
+
+class CallbackContext:
+    """Worker can run multiple Python callbacks, CallbackContext is used to
+    (independently from other callbacks) manage shared memory used to pass
+    results of the callback calls.
+    """
+
+    def __init__(self, callback, mem_chunks):
+        self.callback = callback
+        self.mem_chunks = mem_chunks
+        self.batch_i = 0
+
+    def _next_batch_i(self):
+        res = self.batch_i
+        self.batch_i = (self.batch_i + 1) % len(self.mem_chunks)
+        return res
+
+    def next_mem_chunk(self):
+        batch_i = self._next_batch_i()
+        return self.mem_chunks[batch_i]
```

Comment:
How about:
```suggestion
        while True
            for c in  self.mem_chunks:
                yield c
```

Reply:
I agree that would be nice, but than I need to keep the generator itself and we need to change it to `next(context.mem_chunks_generator)`. I think I prefer what we have now.

Reply:
Sure